src/stan/services/pathfinder/multi.hpp

```diff
@@ -158,53 +169,69 @@ inline int pathfinder_lbfgs_multi(
   for (auto&& ilpr : individual_lp_ratios) {
     num_returned_samples += ilpr.size();
   }
-  Eigen::Array<double, Eigen::Dynamic, 1> lp_ratios(num_returned_samples);
+  // Rows are individual parameters and columns are samples per iteration
   Eigen::Array<double, Eigen::Dynamic, Eigen::Dynamic> samples(
       individual_samples[0].rows(), num_returned_samples);
   Eigen::Index filling_start_row = 0;
   for (size_t i = 0; i < successful_pathfinders; ++i) {
-    const Eigen::Index individ_num_samples = individual_lp_ratios[i].size();
-    lp_ratios.segment(filling_start_row, individ_num_samples)
-        = individual_lp_ratios[i];
+    const Eigen::Index individ_num_samples = individual_samples[i].cols();
     samples.middleCols(filling_start_row, individ_num_samples)
         = individual_samples[i];
     filling_start_row += individ_num_samples;
   }
-  const auto tail_len = std::min(0.2 * num_returned_samples,
-                                 3 * std::sqrt(num_returned_samples));
-  Eigen::Array<double, Eigen::Dynamic, 1> weight_vals
-      = stan::services::psis::psis_weights(lp_ratios, tail_len, logger);
-  boost::ecuyer1988 rng
-      = util::create_rng<boost::ecuyer1988>(random_seed, stride_id);
-  boost::variate_generator<
-      boost::ecuyer1988&,
-      boost::random::discrete_distribution<Eigen::Index, double>>
-      rand_psis_idx(rng,
-                    boost::random::discrete_distribution<Eigen::Index, double>(
-                        boost::iterator_range<double*>(
-                            weight_vals.data(),
-                            weight_vals.data() + weight_vals.size())));
-  for (size_t i = 0; i <= num_multi_draws - 1; ++i) {
-    parameter_writer(samples.col(rand_psis_idx()));
+  double psis_delta_time = 0;
+  if (psis_resample && calculate_lp) {
```

Comment:
Possibly a bit cheeky, but while you're in here would it be much of a pain to move this resampling process to a separate function (possibly living in `psis.hpp`?) which just gets called here? It would be great if I could perform the resampling with multiple existing single-pathfinder results.

I can also open a PR to do it once this one is merged, if you want to keep this one focused on the resampling option, just let me know!

For context, I can't use the TBB for parallelism in [StanEstimators](https://github.com/andrjohns/StanEstimators), so to parallelise multi-pathfinder I'll have to run multiple single-paths and then perform the resampling on the results.

Reply:
Yeah that's reasonable!

Reply:
Actually could we do this in a separate PR? Looking at it rn I'm not sure if you want back just the resampling weights (which would be just calling `psis_weights` or if you want a function that does the resampling and writes to a `parameter_writer`

Reply:
Of course! I'll open a PR once this is in